mbedtls/src/x509/mbedtls_wrappers.rs

```diff
@@ -0,0 +1,241 @@
+/* Copyright (c) Fortanix, Inc.
+ *
+ * Licensed under the GNU General Public License, version 2 <LICENSE-GPL or
+ * https://www.gnu.org/licenses/gpl-2.0.html> or the Apache License, Version
+ * 2.0 <LICENSE-APACHE or http://www.apache.org/licenses/LICENSE-2.0>, at your
+ * option. This file may not be copied, modified, or distributed except
+ * according to those terms. */
+
+#[cfg(not(feature = "std"))]
+use crate::alloc_prelude::*;
+use mbedtls_sys::*;
+
+use crate::private::UnsafeFrom;
+use crate::{forward_mbedtls_calloc, forward_mbedtls_free};
+//use core::ops::{Deref, DerefMut};
+use mbedtls_sys::types::raw_types::c_void;
+use core::ptr::drop_in_place;
+use core::marker::PhantomData;
+use crate::x509::Certificate;
+use crate::error::{Error, Result};
+use core::ptr::NonNull;
+
+#[repr(transparent)]
+pub struct MbedtlsBox<T> {
+    inner: NonNull<T>
+}
+
+#[cfg(feature = "threading")]
+unsafe impl<> Send for MbedtlsBox<Certificate> {}
+
+#[cfg(feature = "threading")]
+unsafe impl<> Sync for MbedtlsBox<Certificate> {}
+
+impl<> MbedtlsBox<Certificate> {
+    pub fn new() -> Result<Self> {
+        unsafe {
+            let inner = forward_mbedtls_calloc(1, core::mem::size_of::<x509_crt>()) as *mut x509_crt;
+            if inner == core::ptr::null_mut() {
+                return Err(Error::SslAllocFailed);
+            }
+
+            x509_crt_init(inner);
+            Ok(Self { inner: NonNull::new_unchecked(inner as *mut Certificate) })
+        }
+    }
+
+    pub fn as_ref(&self) -> &Self {
+        self
+    }
+}
+
+impl<T> Drop for MbedtlsBox<T> {
+    fn drop(&mut self) {
+        unsafe {
+            drop_in_place(self.inner.as_ptr());
+            forward_mbedtls_free(self.inner.as_ptr() as *mut c_void)
+        }
+    }
+}
+
+
+impl Into<*const x509_crt> for &MbedtlsBox<Certificate> {
+    fn into(self) -> *const x509_crt {
+        self.inner.as_ptr() as *const x509_crt
+    }
+}
+
+impl Into<*mut x509_crt> for &mut MbedtlsBox<Certificate> {
+    fn into(self) -> *mut x509_crt {
+        self.inner.as_ptr() as *mut x509_crt
+    }
+}
+
+impl Into<*mut x509_crt> for &MbedtlsBox<Certificate> {
+    fn into(self) -> *mut x509_crt {
+        self.inner.as_ptr() as *mut x509_crt
+    }
+}
+
+
+#[repr(transparent)]
+pub struct MbedtlsList<T> {
+    inner: *mut T,
```

Comment:
```rust
inner: Option<MbedtlsBox<T>>
```

Reply:
Option is an enum and is not transparent - https://doc.rust-lang.org/src/core/option.rs.html#158-167

We break all casting and logic by doing that.